File: Remove-Nth-Node-FromTheEnd/sol.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if(head->next==nullptr)
            return nullptr;

        ListNode* ptr1 = head;
        ListNode* ptr2 = head;
        int i=0;
        int p1=1, p2=1;
        while(ptr1->next!=nullptr){
            i++;
            ptr1=ptr1->next;
            p1++;
            if(i>n){
                ptr2=ptr2->next;
                p2++;
            }
        }
        //cout << p1 << " " << p2 << endl;
        if(ptr2==head && (p1-p2)!=n)
            head=head->next;
        else if(ptr2->next->next==nullptr){
            if(ptr2==head)
                head->next=nullptr;
            else
                ptr2->next=nullptr;
        }
        else
            ptr2->next=ptr2->next->next;
        //cout << ptr1->val << " " << ptr2->val;
        return head;
    }
};

File: Reverse-Nodes-in-k-Groups/sol.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseNode(ListNode ptr1, ListNode ptr2){
        ListNode ptr = ptr1;
        ListNode head = new ListNode(ptr.val, ptr2);
        ptr = ptr.next;

        while(ptr!=ptr2){
            ListNode p = new ListNode(ptr.val, head);
            head = p;
            ptr = ptr.next;
        }
        return head;
    }

    public void printList(ListNode head){
        while(head!=null){
            System.out.print(head.val + " ");
            head = head.next;
        }
        System.out.println("\n");
    }

    public ListNode reverseKGroup(ListNode head, int k) {
        if(k==1)
            return head;

        ListNode follow = head;
        ListNode lead = head;
        int count=1;
        while(count<k){
            lead = lead.next;
            count++;
        }
        if(lead.next==null)
            return reverseNode(head, null);

        head = reverseNode(follow, lead.next);

        ListNode temp = follow.next;
        while(temp.next!=lead.next)
            temp = temp.next;

        lead = temp;
        follow = head;
        lead = lead.next;
        count=1;

        while(lead.next!=null){
            lead = lead.next;
            follow = follow.next;

            count++;
            if(count==k){
                follow.next = reverseNode(follow.next, lead.next);
                count=0;
                temp = follow.next;
                while(temp.next!=lead.next)
                    temp = temp.next;

                lead = temp;
            }
        }

        return head;


    }
}
